cinder/volume/api.py

```diff
@@ -1631,6 +1653,18 @@ def migrate_volume(self, context, volume, host, cluster_name, force_copy,
             LOG.error(msg)
             raise exception.InvalidVolume(reason=msg)
 
+        # Check if there is an affinity/antiaffinity against the volume
+        filter_properties = {}
+        if "scheduler_hint_same_host" in volume.metadata:
+            LOG.debug("Found a scheduler_hint_same_host in volume %s",
+                      volume.metadata["scheduler_hint_same_host"])
+            filter_properties["same_host"] = volume.metadata["scheduler_hint_same_host"]
+
+        if "scheduler_hint_different_host" in volume.metadata:
```

Comment:
This seems to be a pattern.
Can't we then make it more extensible, so that we can easily carry other properties in the future?
Defining a `PERSISTENT_SCHEDULER_HINTS = ['same_host', 'different_host']` and then iterating through this for saving it in the metadata and passing it to the scheduler on migration.

Also, pep8 checks are failing.

Reply:
I would prefer to store all the hints under a single key as json, so this would become:
```python
        if "scheduler_hints" in volume.metadata:
            LOG.debug("Found a schedulers in volume %s",
                      volume.metadata["scheduler_hints"])
            filter_properties.update(json.loads(volume.metadata["scheduler_hints"]))
```

And storing them
```python
        if scheduler_hints:
                metadata["scheduler_hints"] = json.dumps(scheduler_hints)
```

So we do not have to care here which specific hints are persisted.

Reply:
@fwiesel we might not want to store the shard hint for example.

Reply:
True, but I would rather say, we do not want to prohibit migrating the volume to another shard.

Reply:
First off, there aren't other hints that can be passed in according to the cinder API, so there really isn't a need to make this extensible.  I don't want to store json in metadata for cinder.   Nowhere else does cinder do that for volume metadata, unless a user specifies the json value as part of their own defined metadata.

Reply:
> First off, there aren't other hints that can be passed in according to the cinder API, so there really isn't a need to make this extensible.

I read the [spec differently](https://github.com/openstack/cinder/blob/393c2e4ad90c05ebf28cc3a2c65811d7e1e0bc18/cinder/api/schemas/scheduler_hints.py#L26-L76). 
On top of the two you considered, `local_to_instance`, and `query` have been specified.
And my understanding is that the API is intentionally open to extension the choice of setting (` 'additionalProperties': True`).


> I don't want to store json in metadata for cinder. Nowhere else does cinder do that for volume metadata, unless a user specifies the json value as part of their own defined metadata.

I look at the problem from the perspective, how many places to have to change something, if the assumptions here won't hold. And I agree with @leust, it looks like a pattern, and we have now three code sides which encode the knowledge independently, which attributes need to be stored(&loaded) (and how).

I rather see that encapsulated. In Nova dumping a json is the tool of choice. Feel free to solve it differently, but I would say  some degree of encapsulation is required.

After looking it more why cinder isn't doing though, I stumbled though on another problem with the current approach:
Even if you limit it only to the two hints (`different_host`, and `same_host`), they both come in two forms. One is a string, the other is a list, and that needs to get serialised and deserialised somehow.

Now comes another problem: Metadata only allows 255 characters. 36 characters per uuid, and at least one per separator, that's maybe 6 volumes...
I don't think that is a workable limit.